packages/governance/src/ballotBuilder.js

```diff
@@ -21,64 +20,160 @@ const ChoiceMethod = {
   WEIGHT: 'weight',
 };
 
-const buildEqualWeightBallot = (
+/** @type {{
+ *   PARAM_CHANGE: 'param_change',
+ *   ELECTION: 'election',
+ *   SURVEY: 'survey',
+ * }}
+ */
+const ElectionType = {
+  // A parameter is named, and a new value proposed
+  PARAM_CHANGE: 'param_change',
+  // choose one or multiple winners, depending on ChoiceMethod
+  ELECTION: 'election',
+  SURVEY: 'survey',
+};
+
+/** @type {{
+ *   MAJORITY: 'majority',
+ *   NO_QUORUM: 'no_quorum',
+ *   ALL: 'all',
+ * }}
+ */
+const QuorumRule = {
+  MAJORITY: 'majority',
+  NO_QUORUM: 'no_quorum',
+  // The election isn't valid unless all voters cast a ballot
+  ALL: 'all',
+};
+
+const verifyQuestionFormat = (electionType, question) => {
+  switch (electionType) {
+    case ElectionType.SURVEY:
+    case ElectionType.ELECTION:
+      assert.typeof(
+        question.text,
+        'string',
+        X`Question ("${question}") must be a record with text: aString`,
+      );
+      break;
+    case ElectionType.PARAM_CHANGE:
+      assert.typeof(
+        question.paramSpec,
+        'object',
+        X`Question ("${question}") must be a record with paramSpec: anObject`,
+      );
+      assert(question.proposedValue);
+      assertType(ParamType.INSTANCE, question.contract, electionType);
+      break;
+    default:
+      throw Error(`Election type unrecognized`);
+  }
+};
+
+const positionIncluded = (positions, p) =>
+  positions.some(e => sameStructure(e, p));
+
+// BallotSpec contains the subset of BallotDetails that can be specified before
+// the ballot is created.
+const makeBallotSpec = (
   method,
   question,
   positions,
-  maxChoices = 0,
-  instance,
+  electionType,
+  maxChoices,
+  closingRule,
+  quorumRule,
+  tieOutcome,
 ) => {
```

Comment:
Hovering over this in vscode, I see that typescript doesn't know in are or return types.

Experimentally typing it in #3758 revealed a type error that looks like a genuine bug in governParam.js which I comment on there.

I see that it used to have 5 parameters, which is on the edge of what we should encode as positional parameters. (Even 5 is uncomfortable, though I'm sure I've done it.) Eight is definitely too much. But if we encode with name-based properties, then the argument is already a presumed ballotSpec. All this function then does is some sanity checking. In this case it completely ignore several parameters, so it is clearly not doing validation. It seems like another `looksLikeFoo` case is appropriate.

Or, if you really are trying to do input validation, then you should check all inputs that might be malformed.